flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/FileMergingCheckpointUtils.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint.filemerging;
+
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+
+import javax.annotation.Nullable;
+
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+
+/** Utilities for file-merging checkpoints. */
+public class FileMergingCheckpointUtils {
+
+    // ------------------------------------------------------------------------
+    //  FileMergingSnapshotManager initialization related
+    // ------------------------------------------------------------------------
+
+    /** A class that packs the file system info for snapshot. */
+    public static class SnapshotFileSystemInfo {
+        FileSystem fs;
+        Path checkpointBaseDirectory;
+        Path sharedStateDirectory;
+        Path taskOwnedStateDirectory;
+
+        public SnapshotFileSystemInfo(
+                FileSystem fs,
+                Path checkpointBaseDirectory,
+                @Nullable Path sharedStateDirectory,
+                Path taskOwnedStateDirectory) {
+            this.fs = fs;
+            this.checkpointBaseDirectory = checkpointBaseDirectory;
```

Comment:
Need to check if `checkpointBaseDirectory` is prefix of `sharedStateDirectory` and `taskOwnedStateDirectory` here? TBH I am not familiar with it if we plan to relax this restricts.

Reply:
I added checkNotNull here, and I thought we can refer to the initialization of `FsCheckpointStorageLocation`, which does not check the relationship between `checkpointBaseDirectory ` and `sharedStateDirectory ` or `taskOwnedStateDirectory `. The caller is responsible to initialize those directories and ensure the relationship among them.